From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sayakie <sayakie@kakao.com>
Date: Thu, 17 Aug 2023 23:13:14 +0900
Subject: [PATCH] Configurable farmland water seek range


diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index d566f67f8f6f1748023430de4f191881b79e44a1..0d5531b85ffd63039e1d86cc683b7b1fde3c6cd3 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -163,9 +163,9 @@ public class FarmBlock extends Block {
         int yOff = pos.getY();
         int zOff = pos.getZ();
 
-        for (int dz = -4; dz <= 4; ++dz) {
+        for (int dz = -((ServerLevel) world).shuviConfig.farmlandWaterSeekRange; dz <= ((ServerLevel) world).shuviConfig.farmlandWaterSeekRange; ++dz) { // Shuvi
             int z = dz + zOff;
-            for (int dx = -4; dx <= 4; ++dx) {
+            for (int dx = -((ServerLevel) world).shuviConfig.farmlandWaterSeekRange; dx <= ((ServerLevel) world).shuviConfig.farmlandWaterSeekRange; ++dx) { // Shuvi
                 int x = xOff + dx;
                 for (int dy = 0; dy <= 1; ++dy) {
                     int y = dy + yOff;
diff --git a/src/main/java/org/shuvimc/shuvi/ShuviWorldConfig.java b/src/main/java/org/shuvimc/shuvi/ShuviWorldConfig.java
index c6a2e37e348c0c2190e07c2d919bccd8569fc7ac..58d5eceb6a65ab3ce7b27dc130d57577baf0c53a 100644
--- a/src/main/java/org/shuvimc/shuvi/ShuviWorldConfig.java
+++ b/src/main/java/org/shuvimc/shuvi/ShuviWorldConfig.java
@@ -1,7 +1,6 @@
 package org.shuvimc.shuvi;
 
 import org.apache.commons.lang.BooleanUtils;
-import org.bukkit.World;
 import org.bukkit.configuration.ConfigurationSection;
 
 import java.util.List;
@@ -76,4 +75,9 @@ public final class ShuviWorldConfig {
         return value.isEmpty() ? fallback : value;
     }
 
+    public int farmlandWaterSeekRange = 4;
+    private void farmlandSettings() {
+        farmlandWaterSeekRange = getInt("blocks.farmland.water-seek-range", farmlandWaterSeekRange);
+    }
+
 }
